two_factor_otp_auth/models/res_users.py

```diff
@@ -0,0 +1,175 @@
+        string='Two Factor Authentication',
+        inverse="_inverse_enable_2fa",
+    )
+    secret_code_2fa = fields.Char(
+        string='Two Factor Authentication Secret Code',
+        copy=False,
+    )
+    qr_image_2fa = fields.Binary(
+        string='Two Factor Authentication QR Code',
+        copy=False,
+    )
+
+    def _inverse_enable_2fa(self):
+        """
+        Inverse `enable_2fa` - call `action_discard_2f_auth_credentials` method
+        if value of the field become `false`
+        """
+        for user in self:
+            if not user.enable_2fa:
+                user.action_discard_2f_auth_credentials()
+
+    def action_discard_2f_auth_credentials(self):
+        """
+        Remove values from fields `qr_image_2fa`, `auth_secret_code_2fa`
+        """
+        values = {
+            "qr_image_2fa": False,
+            "secret_code_2fa": False,
+        }
+        self.write(values)
+
+    def action_disable_2f_auth(self):
+        """
+        Set `enable_2fa` field value to `False`. Check access for action
+        via `_can_change_2f_auth_settings`.
+        """
+        self._can_change_2f_auth_settings()
+        values = {
+            "enable_2fa": False,
+        }
+        self.write(values)
+
+    def action_enable_2f_auth(self):
+        """
+        Set `enable_2fa` field value to `False`. Check access for action
+        via `_can_change_2f_auth_settings`.
+        """
+        self._can_change_2f_auth_settings()
+        values = {
+            "enable_2fa": True,
+        }
+        self.write(values)
+
+    def _check_credentials(self, password):
+        """
+        Overload core method to also check Two Factor Authentication
+        credentials.
+
+        Raises:
+         * odoo.addons.two_factor_otp_auth.exceptions.MissingOtpError - no
+            `otp_code` in request params. Should be caught by controller and
+            render and open enter "one-time-password" page or QR code creation
+        """
+        self.ensure_one()
+        super(ResUsers, self)._check_credentials(password)
+        if self.enable_2fa:
+            params = request.params
+            secret_code = self.secret_code_2fa or params.get("secret_code_2fa")
+            if params.get("otp_code") is None:
+                request.session.otk_uid = self.id
+                raise MissingOtpError()
+            else:
+                # can trigger `InvalidOtpError`
+                self._check_otp_code(
+                    params.pop("otp_code"),
+                    secret_code,
+                )
+
+    def _generate_secrets(self):
+        """
+        Generate QR-Code based on random set of letters
+
+        Returns:
+         * tuple - generated secret_code and binary qr-code
+        """
+        self.ensure_one()
+        key = b32encode(urandom(10))
+        code = pyotp.totp.TOTP(key).provisioning_uri(self.login)
+        img = qrcode.make(code)
+        _, file_path = mkstemp()  # creating temporary file
```

Comment:
I'm a little worried that you create file, open it and delete it in different blocks. When you work with OS, there can happen different errors. Maybe make sense to do it in try / except / final block? Please, investigate this point.

Reply:
done